README
= Tab Plugin for Redmine

A plugin to allow users to create a new Redmine tab to embed content from an iframe on a per-project base.

== CREDITS ==

- Mischa the Evil for extensive updates and refactoring 
- Eric Davis for identifying a number of issues and providing fixes

== INSTALLATION == 

To install from git: 

$ cd {RAILS_ROOT}/vendor/plugins
$ git clone git://github.com/jamtur01/redmine_tab.git

== INSTRUCTIONS ==

This creates a single tab in which you can embed an iframe. You can globally configure the name of the tab in the Plugins -> Tab Plugin configuration options (default is "Tab Name"). The iframe-links can be configured on a per-project base using the value of a pre-configured project custom field.

1. You need to go to 'Admin -> Custom fields -> Project' and create a field of the type 'Long text' to hold the iframe-link.
2. Select the field you've created in the plugin configuration by selecting Plugins -> Tab Plugin and clicking Configure.
3. Inside your project you will need to add your iframe-link as the value of the field you've selected. For example to embed Google: <iframe src="http://www.google.com/" style="border: 0" width="800" height="600" frameborder="0" scrolling="no"></iframe>
4. Then use the project's Modules tab to enable the Tab module for the project.


CHANGELOG
0.3.1 (2009-08-04)
    * Fixed Redmine trunk compatibility
    * Added Japanese i18n
    * Added Dutch i18n
    * Improved documentation and licensing

0.3.0 (2009-04-11)
    * Fixed tab model duplication
    * Added LICENSE-file
    * Added the ability to define a system "Tab" for the top menu
    * I18n'd the settings page
    * Changed the project menu to allow i18n strings

0.2.0 (2009-02-07)
    * Lots of refactoring
    * Lots of fixes
    * Improved documentation
    * Reintroduced per-project tabs

0.1.0 (2008-08-25)
    - Initial public release

app/models/tab.rb
# tab.rb
# 
# Copyright (C) 2008-2009  James Turnbull <james@lovedthanlost.net>
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

class Tab
  
  # retrieve the value from the configured project custom field
  def self.get_tab_text(project)
       tab_text = project.custom_values.detect {|v| v.custom_field_id == Setting.plugin_redmine_tab['tab_text'].to_i}
       tab_text = tab_text.value if tab_text
  end
end


app/controllers/tab_controller.rb
# tab_controller.rb
#  
# Copyright (C) 2008-2009  James Turnbull <james@lovedthanlost.net>
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

class TabController < ApplicationController
  unloadable
  
  layout 'base'
  
  before_filter :find_project, :authorize, :only => [:show]
  
  def show
    @tab_text = Tab.get_tab_text(@project)
  end

  def system_show
    @tab_text = Setting.plugin_redmine_tab['system_tab_text']
    render :action => 'show'
  end
  
  private
  
  def find_project
  # @project variable must be set before calling the authorize filter
    @project = Project.find(params[:id])
  rescue ActiveRecord::RecordNotFound
    render_404
  end
end


init.rb
# Redmine - project management software
# Copyright (C) 2006-2009  Jean-Philippe Lang
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

require 'redmine'

RAILS_DEFAULT_LOGGER.info 'Starting Tab plugin 0.3.1 for Redmine'

Redmine::Plugin.register :redmine_tab do
  name 'Tab Plugin'
  author 'James Turnbull'
  description 'A plugin which adds Redmine tabs to embed content from an iframe on a per-project and system-wide base.'
  version '0.3.1'
  
  settings :default => {
    'tab_text' => '',
    'tab_name' => 'Tab Name',
    'system_tab_text' => '',
    'system_tab_name' => 'System Tab Name'

  }, :partial => 'settings/redminetab_settings'

  # This plugin adds a project module
  # It can be enabled/disabled at project level (Project settings -> Modules)
  project_module :tab do
    # This permission has to be explicitly given
    # It will be listed on the permissions screen
    permission :view_tab, {:tab => :show}
  end

  # Translate or return the setting_name directly
  string_or_translate = lambda {|setting_name|
    string = Setting.plugin_redmine_tab[setting_name]
    if !string.blank? && string.match(/\A:/) # uses symbol syntax, :string
      string.gsub!(':','')
      string = GLoc.l(string.to_sym) if Object.const_defined?('GLoc') # Rails 2.1.x
      string = I18n.t(string.to_sym) if Object.const_defined?('I18n') # Rails 2.2.x
    end
    string
  }
  
  # A new item is added to the project menu
  menu(:project_menu,
       :tab,
       { :controller => 'tab', :action => 'show' },
       :caption => Proc.new { string_or_translate.call('tab_name') },
       :if => Proc.new { !Setting.plugin_redmine_tab['tab_name'].blank? })

  # A new item is added to the top menu
  menu(:top_menu,
       :tab,
       { :controller => 'tab', :action => 'system_show' },
       :caption => Proc.new { string_or_translate.call('system_tab_name') },
       :if => Proc.new { !Setting.plugin_redmine_tab['system_tab_name'].blank? })
end
